### ISSUE TITLE
ImageDataGenerator.flow() for unlabelled test data (only features X, no labels y)

### ISSUE THREAD

**Issue by asmith26** [fchollet/keras#2694]
For my training data (where I have both X and labels y), I'm creating an `ImageDataGenerator` with `featurewise_center=True` and `featurewise_std_normalization=True`:

```
train_datagen = ImageDataGenerator( featurewise_center=True,
                                    featurewise_std_normalization=True )
train_datagen.fit( X_train )
model.fit_generator( train_datagen.flow(X_train, y_train)
                     samples_per_epoch=X_train.shape[0],
                     nb_epoch=NB_EPOCHS)
```

I am hoping to apply this same image preprocessing to my **unlabelled** test data (this test set came from a Kaggle competition). Following advice from [Stanford cs231n/Karpathy](http://cs231n.github.io/neural-networks-2/) (you can find full details by searching for the term "Common pitfall" on the linked page):

> An important point to make about the preprocessing is that any preprocessing statistics (e.g. the data mean) must only be computed on the training data, and then applied to the validation / test data.

Thus I hoped to use something like:

```
model.predict_generator(train_datagen.flow(X_test),
                        val_samples=X_test.shape[0])
```

however [`ImageDataGenerator.flow()` requires a `y` argument](https://github.com/fchollet/keras/blob/master/keras/preprocessing/image.py#L273). Does anyone know how I might achieve this (I'm using a model built using the functional API)? (Mentioning @chsasank as I see you've made some [great changes to this recently](https://github.com/fchollet/keras/pull/2446) ;) ) Many thanks for any help!

**Comment by chsasank** [comment]
This is a hack which can be used to get it working with current version:

```python
def test_gen(X_test):
    Y_trash = np.ones(X_test.shape[0])
    flow = train_datagen.flow(X_test, Y_trash)
    for X,Y in flow
        yield X #ignore Y
end

model.predict_generator(test_gen(X_test),
                        val_samples=X_test.shape[0])
```

But this is a hack.

We should eventually add following features to `ImageDataGenerator`:
1. Make it serialisable. Currently `pickle.save` doesn't work as `ImageDataGenerator` uses some threading.
2. `flow` should work with 1 or 2 arguments.

**Comment by asmith26** [comment]
Massive thanks for the temporary fix @chsasank!!!!! Works perfectly!! I'll leave this issue open for now in case we wanted to add your more permanent fixes. Thanks again!

**Comment by RaananHadar** [comment]
Hi,
I hope you can improve the documention for ImageDataGenerator. I attempted unsuccessfully to do some pre-processing and resorting to reading the code (which is educational but not preferable to you in the long run...). Many thanks!

**Comment by chsasank** [comment]
It'll be helpful to know what improvements can be made for ImageDataGenerator. I mean it'd be great to know what you found lacking.

In fact, you can raise a PR yourself, now that you have gone through the code! 😛

**Comment by asmith26** [comment]
Thanks @fchollet! Commit https://github.com/fchollet/keras/commit/b71f6ba8644681b51ce4962ad2e0543adbfd4d28 fixes this, hence closing.

**Comment by RaananHadar** [comment]
Hi @chsasank I would love to do so in the future and become seriously more active. I believe in this project. Sadly currently I am head over heels with my thesis. I hope some of you can reciprocate :)

Here is a [link to an issue](https://github.com/asmith26/wide_resnets_keras/issues/1) I opened in a keras code repository @asmith26  wrote a few days ago. It attempts to use ZCA the way I thought you would use it but reaches unsatisfying results. I would love to hear your insight on the manner.

Thanks!
